In [3]:
import pandas as pd
import numpy as np

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [4]:
#Create a DataFrame
d = {
    'Name':['Alisa','Bobby','jodha','jack','raghu','Cathrine',
            'Alisa','Bobby','kumar','Alisa','Alex','Cathrine'],
    'Age':[26,24,23,22,23,24,26,24,22,23,24,24],
      
    'Score':[85,63,55,74,31,77,85,63,42,62,89,77]}
 
df = pd.DataFrame(d,columns=['Name','Age','Score'])
df

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
6      Alisa   26     85
7      Bobby   24     63
8      kumar   22     42
9      Alisa   23     62
10      Alex   24     89
11  Cathrine   24     77

# Duplicates

#### Find the duplicate row in pandas:
duplicated() function is used for find the duplicate rows of the dataframe in python pandas

In [9]:
df.duplicated()
 

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9     False
10    False
11     True
dtype: bool

The above code finds whether the row is duplicate and tags TRUE if it is duplicate and tags FALSE if it is not duplicate. 

And assigns it to the column named “is_duplicate”  of the dataframe  df.

In [6]:
df.duplicated('Name')


0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9      True
10    False
11     True
dtype: bool

This checks if there are duplicate values in a particular column of your DataFrame.

#### Getting rid of duplicates

In [10]:
df.drop_duplicates()

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
8      kumar   22     42
9      Alisa   23     62
10      Alex   24     89

In [11]:
# Sometimes, you may want to drop duplicates just from one column.

df.drop_duplicates(['Name'])

Name  Age  Score
0      Alisa   26     85
1      Bobby   24     63
2      jodha   23     55
3       jack   22     74
4      raghu   23     31
5   Cathrine   24     77
8      kumar   22     42
10      Alex   24     89

# checking basic sanity of data

In [42]:
from sklearn import datasets

In [43]:
iris = datasets.load_iris()

# or

# iris = pd.read_csv("F:\SimpliLearn\ML Certification Course\Helping Documents\DATASETS\iris.csv")
# iris.head()

In [44]:
X = iris.data
y = iris.target

In [45]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [47]:
X[:5]

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2],
       [ 5. ,  3.6,  1.4,  0.2]])

In [48]:
y[:5]

array([0, 0, 0, 0, 0])

In [49]:
np.c_[X[:5], y[:5]]

array([[ 5.1,  3.5,  1.4,  0.2,  0. ],
       [ 4.9,  3. ,  1.4,  0.2,  0. ],
       [ 4.7,  3.2,  1.3,  0.2,  0. ],
       [ 4.6,  3.1,  1.5,  0.2,  0. ],
       [ 5. ,  3.6,  1.4,  0.2,  0. ]])

In [50]:
data    = np.c_[X, y]
columns = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'class']

df      = pd.DataFrame(data, columns=columns)
df.head()

sepal_length  sepal_width  petal_length  petal_width  class
0           5.1          3.5           1.4          0.2    0.0
1           4.9          3.0           1.4          0.2    0.0
2           4.7          3.2           1.3          0.2    0.0
3           4.6          3.1           1.5          0.2    0.0
4           5.0          3.6           1.4          0.2    0.0

In [51]:
# change the class from numeric to string for easy understanding
replacements  = {0: 'setosa', 1: 'versicolor', 2: 'virginica'}
df['class'] = df['class'].map(replacements)

In [53]:
# find out how many samples from each class
df['class'].value_counts(normalize=True)

# bankdata['account_type'].value_counts(normalize=True)
# saving 80%
# current 18%
# personal 2%

setosa        0.333333
versicolor    0.333333
virginica     0.333333
Name: class, dtype: float64

In [31]:
# basic statistics
# check for min, max, quartiles, IQR
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [56]:
df[['sepal_length', 'petal_length']].describe()

sepal_length  petal_length
count    150.000000    150.000000
mean       5.843333      3.758000
std        0.828066      1.765298
min        4.300000      1.000000
25%        5.100000      1.600000
50%        5.800000      4.350000
75%        6.400000      5.100000
max        7.900000      6.900000

In [58]:
# calculate the mean petal length of each of the three classes
# df.loc[df['class'] == 'setosa', metrics].mean(axis=0)

In [31]:
# returns indexes
# df['class'] == 'setosa'

In [59]:
# locate by index position
df.loc[[0, 6]]

sepal_length  sepal_width  petal_length  petal_width   class
0           5.1          3.5           1.4          0.2  setosa
6           4.6          3.4           1.4          0.3  setosa

In [60]:
df.loc[df['class'] == 'setosa']

sepal_length  sepal_width  petal_length  petal_width   class
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8          4.0           1.2          0.2  setosa
15           5.7          4.4           1.5          0.4  setosa
16           5.4          3.9           1.3          0.4  setosa
17           5.1          3.5           1.4          0.3  setosa
18           5.7          3.8           1.7          0.3  setosa
19           5.1          3.8           1.5          0.3  setosa
20           5.4          3.4           1.7          0.2  setosa
21           5.1          3.7           1.5          0.4  setosa
22           4.6          3.6           1.0          0.2  setosa
23           5.1          3.3           1.7          0.5  setosa
24           4.8          3.4           1.9          0.2  setosa
25           5.0          3.0           1.6          0.2  setosa
26           5.0          3.4           1.6          0.4  setosa
27           5.2          3.5           1.5          0.2  setosa
28           5.2          3.4           1.4          0.2  setosa
29           4.7          3.2           1.6          0.2  setosa
30           4.8          3.1           1.6          0.2  setosa
31           5.4          3.4           1.5          0.4  setosa
32           5.2          4.1           1.5          0.1  setosa
33           5.5          4.2           1.4          0.2  setosa
34           4.9          3.1           1.5          0.2  setosa
35           5.0          3.2           1.2          0.2  setosa
36           5.5          3.5           1.3          0.2  setosa
37           4.9          3.6           1.4          0.1  setosa
38           4.4          3.0           1.3          0.2  setosa
39           5.1          3.4           1.5          0.2  setosa
40           5.0          3.5           1.3          0.3  setosa
41           4.5          2.3           1.3          0.3  setosa
42           4.4          3.2           1.3          0.2  setosa
43           5.0          3.5           1.6          0.6  setosa
44           5.1          3.8           1.9          0.4  setosa
45           4.8          3.0           1.4          0.3  setosa
46           5.1          3.8           1.6          0.2  setosa
47           4.6          3.2           1.4          0.2  setosa
48           5.3          3.7           1.5          0.2  setosa
49           5.0          3.3           1.4          0.2  setosa

In [61]:
# display selected columns for class = 'setosa'
df.loc[df['class'] == 'setosa', ['petal_width', 'petal_length']]

petal_width  petal_length
0           0.2           1.4
1           0.2           1.4
2           0.2           1.3
3           0.2           1.5
4           0.2           1.4
5           0.4           1.7
6           0.3           1.4
7           0.2           1.5
8           0.2           1.4
9           0.1           1.5
10          0.2           1.5
11          0.2           1.6
12          0.1           1.4
13          0.1           1.1
14          0.2           1.2
15          0.4           1.5
16          0.4           1.3
17          0.3           1.4
18          0.3           1.7
19          0.3           1.5
20          0.2           1.7
21          0.4           1.5
22          0.2           1.0
23          0.5           1.7
24          0.2           1.9
25          0.2           1.6
26          0.4           1.6
27          0.2           1.5
28          0.2           1.4
29          0.2           1.6
30          0.2           1.6
31          0.4           1.5
32          0.1           1.5
33          0.2           1.4
34          0.2           1.5
35          0.2           1.2
36          0.2           1.3
37          0.1           1.4
38          0.2           1.3
39          0.2           1.5
40          0.3           1.3
41          0.3           1.3
42          0.2           1.3
43          0.6           1.6
44          0.4           1.9
45          0.3           1.4
46          0.2           1.6
47          0.2           1.4
48          0.2           1.5
49          0.2           1.4

In [62]:
# display mean values of selected columns for class = 'setosa'
df.loc[df['class'] == 'setosa', ['petal_width', 'petal_length']].mean()

petal_width     0.246
petal_length    1.462
dtype: float64

In [46]:
# display mean values of selected columns for class = 'setosa'
df.loc[df['class'] == 'setosa', ['petal_width', 'petal_length']].max()

petal_width     0.6
petal_length    1.9
dtype: float64

In [63]:
df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [64]:
# use of grouping
groups = df.groupby('class')

In [65]:
# use sum(), min(), max(), count()
groups.mean()

sepal_length  sepal_width  petal_length  petal_width
class                                                           
setosa             5.006        3.428         1.462        0.246
versicolor         5.936        2.770         4.260        1.326
virginica          6.588        2.974         5.552        2.026

In [66]:
groups.agg(['mean', 'min', 'max', 'count'])

sepal_length                 sepal_width                  \
                   mean  min  max count        mean  min  max count   
class                                                                 
setosa            5.006  4.3  5.8    50       3.428  2.3  4.4    50   
versicolor        5.936  4.9  7.0    50       2.770  2.0  3.4    50   
virginica         6.588  4.9  7.9    50       2.974  2.2  3.8    50   

           petal_length                 petal_width                  
                   mean  min  max count        mean  min  max count  
class                                                                
setosa            1.462  1.0  1.9    50       0.246  0.1  0.6    50  
versicolor        4.260  3.0  5.1    50       1.326  1.0  1.8    50  
virginica         5.552  4.5  6.9    50       2.026  1.4  2.5    50

# numeric col containing non-numeric data

In [32]:
d = {
 'unit': ['UD', 'UD', 'UD', 'UD', 'ud','uD'],
 'N-D': [ 'Q1', 'Q2', 'Q3', 'Q4','Q5','Q6'],
 'num' : [ -1.48, 1.7, -6.18, 0.25, 'weird', 0.25]

}
df = pd.DataFrame(d)
df

unit N-D    num
0   UD  Q1  -1.48
1   UD  Q2    1.7
2   UD  Q3  -6.18
3   UD  Q4   0.25
4   ud  Q5  weird
5   uD  Q6   0.25

In [35]:
df['num'].str.isnumeric()

0      NaN
1      NaN
2      NaN
3      NaN
4    False
5      NaN
Name: num, dtype: object

In [36]:
df['num'].str.isdecimal()

0      NaN
1      NaN
2      NaN
3      NaN
4    False
5      NaN
Name: num, dtype: object

# check for character cases

In [37]:
df['unit'].str.isupper()

0     True
1     True
2     True
3     True
4    False
5    False
Name: unit, dtype: bool

In [71]:
df['unit'] = df['unit'].str.upper()

In [66]:
df

N-D    num unit
0  Q1  -1.48   UD
1  Q2    1.7   UD
2  Q3  -6.18   UD
3  Q4   0.25   UD
4  Q5  weird   UD
5  Q6   0.25   UD

# handling missing values

In [40]:
location = "F:\SimpliLearn\ML Certification Course\Helping Documents\DATASETS\nba.csv"

In [41]:
# making data frame from csv file 
nba = pd.read_csv(location) 

FileNotFoundError: File b'F:\\SimpliLearn\\ML Certification Course\\Helping Documents\\DATASETS\nba.csv' does not exist

In [75]:
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [76]:
# are there nulls?
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary      11
dtype: int64

example #1 - Replacing NaN values with a Static value.

In [84]:
# replacing na values in college with No college 
nba["College"].fillna("No College", inplace = True) 
nba.isnull().sum()
#nba['College'].isnull().sum()

Name        0
Team        0
Number      0
Position    0
Age         0
Height      0
Weight      0
College     0
Salary      0
dtype: int64

Example #2: Using method Parameter

In [79]:
# making data frame from csv file 
nba = pd.read_csv(location) 
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [78]:
# making data frame from csv file 
nba = pd.read_csv(location) 
  
# replacing na values in college with No college 
nba["College"].fillna( method ='ffill', inplace = True) 
  
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4      Georgia State   5000000.0  
5      Georgia State  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

Similarly, bfill, backfill and pad methods can also be used.

Example #3: Using mean value

In [80]:
# making data frame from csv file 
nba = pd.read_csv(location) 

In [81]:
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary      11
dtype: int64

In [82]:
# Or, depending on your needs:
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].mean())
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].min())
nba['Salary'] = nba['Salary'].fillna(nba['Salary'].max())

In [83]:
nba.isnull().sum()

Name         0
Team         0
Number       0
Position     0
Age          0
Height       0
Weight       0
College     84
Salary       0
dtype: int64

In [100]:
nba['Age'] = nba['Age'].fillna(200)

In [101]:
nba[nba['Age']==200]

Empty DataFrame
Columns: [Name, Team, Number, Position, Age, Height, Weight, College, Salary]
Index: []

In [85]:
raw_data = {'first_name': ['Jason', np.nan, 'Tina', 'Jake', 'Amy'], 
        'last_name': ['Miller', np.nan, 'Ali', 'Milner', 'Cooze'], 
        'age': [42, np.nan, 36, 24, 73], 
        'sex': ['m', np.nan, 'f', 'm', 'f'], 
        'preTestScore': [4, np.nan, np.nan, 2, 3],
        'postTestScore': [25, np.nan, np.nan, 62, 70]}
df = pd.DataFrame(raw_data, columns = ['first_name', 'last_name', 'age', 'sex', 'preTestScore', 'postTestScore'])
df

first_name last_name   age  sex  preTestScore  postTestScore
0      Jason    Miller  42.0    m           4.0           25.0
1        NaN       NaN   NaN  NaN           NaN            NaN
2       Tina       Ali  36.0    f           NaN            NaN
3       Jake    Milner  24.0    m           2.0           62.0
4        Amy     Cooze  73.0    f           3.0           70.0

In [86]:
# Drop missing observations
# rows returned after dropping, the original df remains intact
df_no_missing = df.dropna()
df_no_missing

first_name last_name   age sex  preTestScore  postTestScore
0      Jason    Miller  42.0   m           4.0           25.0
3       Jake    Milner  24.0   m           2.0           62.0
4        Amy     Cooze  73.0   f           3.0           70.0

In [87]:
#Drop rows where all cells in that row is NA
df_cleaned = df.dropna(how='all')
df_cleaned

first_name last_name   age sex  preTestScore  postTestScore
0      Jason    Miller  42.0   m           4.0           25.0
2       Tina       Ali  36.0   f           NaN            NaN
3       Jake    Milner  24.0   m           2.0           62.0
4        Amy     Cooze  73.0   f           3.0           70.0

In [88]:
# Create a new column full of missing values
df['location'] = np.nan
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           NaN            NaN       NaN
2       Tina       Ali  36.0    f           NaN            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [89]:
# Drop column if they only contain missing values
df.dropna(axis=1, how='all')

first_name last_name   age  sex  preTestScore  postTestScore
0      Jason    Miller  42.0    m           4.0           25.0
1        NaN       NaN   NaN  NaN           NaN            NaN
2       Tina       Ali  36.0    f           NaN            NaN
3       Jake    Milner  24.0    m           2.0           62.0
4        Amy     Cooze  73.0    f           3.0           70.0

In [109]:
# Drop rows that contain less than five observations
# This is really mostly useful for time series
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           NaN            NaN       NaN
2       Tina       Ali  36.0    f           NaN            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [110]:
df.dropna(thresh=5)

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       NaN
3       Jake    Milner  24.0   m           2.0           62.0       NaN
4        Amy     Cooze  73.0   f           3.0           70.0       NaN

In [90]:
# Fill in missing data with zeros
df.fillna(0)

first_name last_name   age sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0   m           4.0           25.0       0.0
1          0         0   0.0   0           0.0            0.0       0.0
2       Tina       Ali  36.0   f           0.0            0.0       0.0
3       Jake    Milner  24.0   m           2.0           62.0       0.0
4        Amy     Cooze  73.0   f           3.0           70.0       0.0

In [112]:
# Fill in missing in preTestScore with the mean value of preTestScore
# inplace=True means that the changes are saved to the df right away

In [91]:
df["preTestScore"].fillna(df["preTestScore"].mean(), inplace=True)
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           3.0            NaN       NaN
2       Tina       Ali  36.0    f           3.0            NaN       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

In [115]:
# Fill in missing in postTestScore with each sex’s mean value of postTestScore
df["postTestScore"].fillna(df.groupby("sex")["postTestScore"].transform("mean"), inplace=True)
df

first_name last_name   age  sex  preTestScore  postTestScore  location
0      Jason    Miller  42.0    m           4.0           25.0       NaN
1        NaN       NaN   NaN  NaN           3.0            NaN       NaN
2       Tina       Ali  36.0    f           3.0           70.0       NaN
3       Jake    Milner  24.0    m           2.0           62.0       NaN
4        Amy     Cooze  73.0    f           3.0           70.0       NaN

#### using scikit learn imputer

strategy : string, optional (default=”mean”)
The imputation strategy.

- If “mean”, then replace missing values using the mean along each column. Can only be used with numeric data.
- If “median”, then replace missing values using the median along each column. Can only be used with numeric data.
- If “most_frequent”, then replace missing using the most frequent value along each column. Can be used with strings or numeric data.
- If “constant”, then replace missing values with fill_value. Can be used with strings or numeric data.

In [97]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import Imputer

df = pd.DataFrame([[10,     20],
                   [np.nan, 23],
                   [15,    np.nan],
                   [17,    29]])

df.dtypes

0    float64
1    float64
dtype: object

In [99]:
imp = SimpleImputer(strategy="median")

print(imp.fit_transform(df)) 

[[10. 20.]
 [15. 23.]
 [15. 23.]
 [17. 29.]]


one more example ...

In [16]:
X = np.array([[23.56],
              [53.45],
              ['NaN'],
              [44.44],
              [77.78],
              ['NaN'],
              [234.44],
              [11.33],
              [79.87]])

In [17]:
imputer = SimpleImputer(strategy='mean')  

imputer.fit_transform(X)

array([[ 23.56      ],
       [ 53.45      ],
       [ 74.98142857],
       [ 44.44      ],
       [ 77.78      ],
       [ 74.98142857],
       [234.44      ],
       [ 11.33      ],
       [ 79.87      ]])

one more example ..

In [18]:
# Create an empty dataset
df = pd.DataFrame()

# Create two variables called x0 and x1. Make the first value of x1 a missing value
df['x0'] = [0.3051,0.4949,0.6974,0.3769,0.2231,0.341,0.4436,0.5897,0.6308,0.5]
df['x1'] = [np.nan,0.2654,0.2615,0.5846,0.4615,0.8308,0.4962,0.3269,0.5346,0.6731]

# View the dataset
df

x0      x1
0  0.3051     NaN
1  0.4949  0.2654
2  0.6974  0.2615
3  0.3769  0.5846
4  0.2231  0.4615
5  0.3410  0.8308
6  0.4436  0.4962
7  0.5897  0.3269
8  0.6308  0.5346
9  0.5000  0.6731

In [19]:
# Fit Imputer
# Create an imputer object that looks for 'Nan' values, then replaces them with the mean value of the feature by columns (axis=0)
mean_imputer = Imputer(missing_values='NaN', strategy='mean', axis=0)

# Train the imputor on the df dataset
mean_imputer = mean_imputer.fit(df)

C:\Anaconda3\lib\site-packages\sklearn\utils\deprecation.py:58: DeprecationWarning: Class Imputer is deprecated; Imputer was deprecated in version 0.20 and will be removed in 0.22. Import impute.SimpleImputer from sklearn instead.
  warnings.warn(msg, category=DeprecationWarning)


In [20]:
# Apply Imputer
# Apply the imputer to the df dataset
imputed_df = mean_imputer.transform(df.values)

In [21]:
imputed_df

array([[0.3051    , 0.49273333],
       [0.4949    , 0.2654    ],
       [0.6974    , 0.2615    ],
       [0.3769    , 0.5846    ],
       [0.2231    , 0.4615    ],
       [0.341     , 0.8308    ],
       [0.4436    , 0.4962    ],
       [0.5897    , 0.3269    ],
       [0.6308    , 0.5346    ],
       [0.5       , 0.6731    ]])

Notice that 0.49273333 is the imputed value, replacing the np.NaN value.

#### more examples ... multiple columns

In [23]:
titanic = pd.read_csv("https://raw.githubusercontent.com/amueller/scipy-2017-sklearn/master/notebooks/datasets/titanic3.csv")
# there is still a small problem with using the OneHotEncoder and missing values,
# so for now I am going to assume there are no missing values by dropping them

titanic2 = titanic.dropna(subset=['pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked'])

In [24]:
target   = titanic2.survived.values
features = titanic2[['pclass', 'sex', 'age', 'fare', 'embarked']]

In [25]:
features.head()

pclass     sex      age      fare embarked
0       1  female  29.0000  211.3375        S
1       1    male   0.9167  151.5500        S
2       1  female   2.0000  151.5500        S
3       1    male  30.0000  151.5500        S
4       1  female  25.0000  151.5500        S

In [34]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer, make_column_transformer

In [35]:
preprocess = make_column_transformer(
    (StandardScaler(), ['age', 'fare']),
    (OneHotEncoder(),  ['pclass', 'sex', 'embarked'])
)

In [39]:
new_features = preprocess.fit_transform(features)
new_features[:3]

array([[-0.05663194,  3.13554913,  1.        ,  0.        ,  0.        ,
         1.        ,  0.        ,  0.        ,  0.        ,  1.        ],
       [-2.01237899,  2.06268333,  1.        ,  0.        ,  0.        ,
         0.        ,  1.        ,  0.        ,  0.        ,  1.        ],
       [-1.93693697,  2.06268333,  1.        ,  0.        ,  0.        ,
         1.        ,  0.        ,  0.        ,  0.        ,  1.        ]])

In [40]:
new_features_df = pd.DataFrame(data=new_features)
new_features_df.head()

0         1    2    3    4    5    6    7    8    9
0 -0.056632  3.135549  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0
1 -2.012379  2.062683  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0
2 -1.936937  2.062683  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0
3  0.013009  2.062683  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0
4 -0.335196  2.062683  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0